### PR TITLE
SimpleWeatherService: Generate random weather data

### PR DESCRIPTION
**Pull request by vkareh**
Generates random weather and forecast data. Also adds a new simulation key handler for weather data:
- 'w' will generate new weather, cycling through the various conditions
- 'W' will clear the weather data

Needed to build the sim for the weather app in https://github.com/InfiniTimeOrg/InfiniTime/pull/1995

### CODE REVIEW

**Comment by vkareh** [comment]
@NeroBurner yes, that was my first iteration, but had issues refreshing the weather since the instance of weatherService in `main.cpp` was different than the instance being used by the displayapp. I've finally solved it and now you can use `w` for generating random weather and `W` for clearing it.

**Comment by vkareh** [comment]
Adding weather to the sim exposed a bug in the PTS face, where the value will always be considered dirty even if the weather data is cleared. I've submitted a fix here: https://github.com/InfiniTimeOrg/InfiniTime/pull/1965

**Comment by vkareh** [comment]
I also added forecast data now.

File: sim/components/ble/SimpleWeatherService.h [hunk 1]
```diff
@@ -80,11 +80,18 @@ class SimpleWeatherService {
       int16_t minTemperature;
       int16_t maxTemperature;
       Icons iconId;
+
+      bool operator==(const Day& other) const;
```
Comment (NeroBurner) [inline sim/components/ble/SimpleWeatherService.h:84]: where does this function come from? I don't think it is needed for the current implementation, and it is not in the current InfiniTime version of `SimpleWeatherService.h`

File: sim/components/ble/SimpleWeatherService.h [hunk 2]
```diff
@@ -80,11 +80,18 @@ class SimpleWeatherService {
...
     };
 
     std::array<Day, MaxNbForecastDays> days;
+
+    bool operator==(const Forecast& other) const;
```
Comment (NeroBurner) [inline sim/components/ble/SimpleWeatherService.h:89]: don't think needed

File: sim/components/ble/SimpleWeatherService.cpp [hunk 3]
```diff
@@ -108,3 +122,17 @@ bool SimpleWeatherService::CurrentWeather::operator==(const SimpleWeatherService
...
+}
+
+bool SimpleWeatherService::Forecast::operator==(const SimpleWeatherService::Forecast& other) const {
+  for (int i = 0; i < this->nbDays; i++) {
+    if (this->days[i] != other.days[i]) {
+      return false;
+    }
+  }
+  return this->timestamp == other.timestamp && this->nbDays == other.nbDays;
+}
```
Comment (NeroBurner) [inline sim/components/ble/SimpleWeatherService.cpp:138]: don't needed (I think)

File: sim/components/ble/SimpleWeatherService.cpp [hunk 4]
```diff
@@ -70,6 +70,20 @@ void SimpleWeatherService::Init() {
   //ble_gatts_add_svcs(serviceDefinition);
 }
 
+void SimpleWeatherService::SetCurrentWeather(uint64_t timestamp, int16_t temperature, int iconId) {
+  SimpleWeatherService::Location cityName = (SimpleWeatherService::Location)"New York";
```
Comment (NeroBurner) [inline sim/components/ble/SimpleWeatherService.cpp:74]: non-blocking: doesn't look like a clean solution, but I think it works. Probably no `\0' after the last character, only at the very last entry of the character array

**Review by NeroBurner — APPROVED** [review]
sorry for the long delay to do this review. Thank you looks perfect! Just some extra code I suspect is a leftover of refactoring. But works just as I wanted. Perfect!

Please inform me if the extra code should stay in (and why). After that I'll happily merge :)

File: sim/components/ble/SimpleWeatherService.cpp (hunk 4 above)
Comment (vkareh) [reply sim/components/ble/SimpleWeatherService.cpp:74]: Yeah I don't like this solution very much. I don't think there's currently a use for it either. I can remove it.

**Comment by vkareh** [comment]
@NeroBurner 
> Please inform me if the extra code should stay in (and why). After that I'll happily merge :)

Absolutely! Thanks for the review. The extra code is just operator overrides to support the Forecast/Day structs. Without it, any screen that relies on `DirtyValue` to display up-to-date Forecast data will make the sim build fail.

I currently added it to the InfiniTime SimpleWeatherService in https://github.com/InfiniTimeOrg/InfiniTime/pull/1995, but I can submit it as a separate PR to add just the overrides if you prefer.

**Comment by vkareh** [comment]
I've submitted a patch for InfiniTime with _only_ the operator overrides: https://github.com/InfiniTimeOrg/InfiniTime/pull/2011

File: sim/components/ble/SimpleWeatherService.cpp (hunk 4 above)
Comment (vkareh) [reply sim/components/ble/SimpleWeatherService.cpp]: Removed this, as it's unused/untested currently.

**Comment by NeroBurner** [comment]
That makes sense! Then yes please keep the operator overrides. They are indeed useful and I am positive they will find their way into InfiniTime eventually 😁

File: sim/components/ble/SimpleWeatherService.h (hunk 2 above)
Comment (NeroBurner) [reply sim/components/ble/SimpleWeatherService.h:89]: Needed for future forecasting app

File: sim/components/ble/SimpleWeatherService.cpp (hunk 3 above)
Comment (NeroBurner) [reply sim/components/ble/SimpleWeatherService.cpp:138]: Needed for future forecasting app

**Comment by NeroBurner** [comment]
One VERY last thing. Please update the readme.md as well with the new key to show and clear weather data 🙇‍♂️

**Comment by vkareh** [comment]
@NeroBurner 
> One VERY last thing. Please update the readme.md as well with the new key to show and clear weather data 🙇‍♂️

Done!

**Comment by NeroBurner** [comment]
Merged! Thanks a lot!

**Pull request merged**